tests/js/tags/admin.js

```diff
@@ -0,0 +1,498 @@
+
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import ReactTable from 'react-table';
+
+import {shallow} from 'enzyme';
+
+import {Container} from 'widgets/columns';
+
+import {ajax} from 'utils/ajax';
+
+import {TagResourcesButton, TagResourcesLoader} from 'tags/admin/';
+
+import TagResourceManager from 'tags/admin/manager';
+import TagResourceSearch from 'tags/admin/search';
+import TagResourceTable from 'tags/admin/table';
+
+
+test('TagResourcesButton render', () => {
+
+    class MockTagResourcesButton extends TagResourcesButton {
+
+        renderButton () {
+            return 7
+        }
+
+        renderResourceManager () {
+            return 23
+        }
+    }
+    const button = shallow(<MockTagResourcesButton />);
+    expect(button.text()).toBe("723");
+});
+
+
+test('TagResourcesButton renderButton', () => {
+    const button = shallow(<TagResourcesButton />);
+    const evt = {preventDefault: jest.fn()};
+
+    expect(button.state().open).toBe(false)
+    button.instance().handleClick(evt);
+    expect(button.state().open).toBe(true)
+    expect(evt.preventDefault.mock.calls).toEqual([[]])
+    button.instance().handleClick(evt);
+    expect(button.state().open).toBe(false)
+    expect(evt.preventDefault.mock.calls).toEqual([[], []])
+})
+
+
+test('TagResourcesButton renderButton', () => {
+    const button = shallow(<TagResourcesButton />);
+    let result = button.instance().renderButton();
+    expect(result.props.children).toBe("Manage resources for this tag");
+    expect(result.props.onClick).toBe(button.instance().handleClick);
+    button.setState({open: true});
+    result = button.instance().renderButton();
+    expect(result.props.children).toBe("Hide the resource manager for this tag");
+    expect(result.props.onClick).toBe(button.instance().handleClick);
+});
+
+
+test('TagResourcesButton renderResourceManager', () => {
+    const button = shallow(
+        <TagResourcesButton
+           tag='foo'
+           api='bar'
+           project='baz' />);
+    let result = button.instance().renderResourceManager();
+    expect(result).toBe('');
+    button.setState({open: true});
+    result = button.instance().renderResourceManager();
+    expect(result.props.tag).toBe('foo');
+    expect(result.props.api).toBe('bar');
+    expect(result.props.project).toBe('baz');
+});
+
+
+test('TagResourceManager render', () => {
+    const _refreshData = jest.fn();
+
+    class MockTagResourceManager extends TagResourceManager {
+
+        async refreshData () {
+            _refreshData();
+        }
+    }
+
+    const manager = shallow(<MockTagResourceManager />);
+    expect(manager.text()).toBe("<TagResourceSearch /><ErrorList /><TagResourceTable />");
+    expect(_refreshData.mock.calls).toEqual([[]]);
+});
```

Comment:
> I am also very unsure about testing the output of the render method. In django, it would like testing that a view returns the correct HTML. I can see myself doing that, but only for very specific cases where the markup is super important.

mostly its doing `shallow` renders - this is the accepted way to test component rendering without actually rendering, i believe.

in some cases, it wants to test specific functionality - and these  tests use `mount` rather than `shallow`